src/Polyhedron.cuh

```diff
@@ -44,4 +45,54 @@ public:
 };
 
 
+/**
+ * Returns the coordinates of the end of AB vector's overlay on the polyhedron's surface
+ *
+ * If vector AB completely lies on the current face of polyhedron, coordinates of point B are returned
+ * If vector AB crosses the edge of current face, the end of overlay on the next face to edge is returned
+ *
+ * @param a The beginning of vector AB, point A
+ * @param b The end of vector AB, point B
+ * @param current_face_id Identifier of the face point A belongs to
+ * @param polyhedron The polyhedron in simulation
+ *
+ * @returns Coordinates of the end of AB vector's overlay
+ */
+__device__ SpacePoint get_projected_vector_end(SpacePoint a, SpacePoint b, int current_face_id, Polyhedron *polyhedron);
+
+/**
+ * Returns the intersection point of segment AB with an edge of given face if it exists, point B otherwise
+ *
+ * @param a Point A of segment AB
+ * @param b Point B of segment AB
+ * @param current_face Face the segment AB belongs to
+ *
+ * @returns The point of intersection with edge if it exists, point B otherwise
+ */
+__device__ SpacePoint find_intersection_with_edge(SpacePoint a, SpacePoint b, Face *current_face);
+
+/**
+ * Finds a face adjacent to the given face along the edge AB
+ *
+ * @param a Point A of edge AB
+ * @param b Point B of edge AB
+ * @param current_face_id Identifier of given face
+ * @param polyhedron The polyhedron in simulation
+ *
+ * @returns Identifier of the found face or `first_face_id` if nothing was found
+ */
+__device__ int find_face_next_to_edge(SpacePoint a, SpacePoint b, int current_face_id, Polyhedron *polyhedron);
```

Comment:
You should either replace first two arguments with something like a vertex id (which will represent an edge with points `polyhedron->faces[current_face_id].vertices[vertex_id]` and `polyhedron->faces[current_face_id].vertices[vertex_id + 1]`), or document the behaviour if AB is not an edge of the given face